cdap-ui/app/services/data/datasource.js

```diff
@@ -115,54 +164,129 @@ angular.module(PKG.name+'.services')
      * poll a resource
      */
     DataSource.prototype.poll = function (resource, cb) {
-      this.bindings.push({
-        poll: true,
-        resource: resource,
-        callback: cb
-      });
+      var self = this;
+      var prom = new MyPromise(function(resolve, reject) {
+        var generatedResource = {};
+        if (!resource.url) {
+          generatedResource = resource;
+        } else {
+          generatedResource = {
+            url: buildUrl(resource.url, resource.params || {}),
+            json: true,
+            method: resource.method
+          };
+        }
 
-      this.scope.$on('$destroy', function () {
-        _pollStop(resource);
-      });
+        self.bindings.push({
+          poll: true,
+          resource: generatedResource,
+          callback: cb,
+          resolve: resolve,
+          reject: reject
+        });
+        self.scope.$on('$destroy', function () {
+          _pollStop(generatedResource);
+        });
 
-      _pollStart(resource);
+        _pollStart(generatedResource);
+      }, true);
+      return prom;
     };
 
-
+    DataSource.prototype.pollStop = function(resource) {
+      var prom = new MyPromise(function(resolve, reject) {
+        _pollStop(resource);
+      });
+      return prom;
+    };
 
 
     /**
      * fetch a resource
      */
     DataSource.prototype.request = function (resource, cb) {
-      var once = false,
-          deferred = $q.defer();
-
-      this.bindings.push({
-        resource: resource,
-        callback: function (result) {
-          if(!once) {
-            once = true;
-            /*jshint -W030 */
-            cb && cb.apply(this, arguments);
-            deferred.resolve(result);
-          }
-        },
-        errorCallback: function (err) {
-          if(!once) {
-            once = true;
-            deferred.reject(err);
+      var self = this;
+      var prom = new MyPromise(function(resolve, reject) {
+
+        var generatedResource = {};
+        if (!resource.url) {
+          generatedResource = resource;
+        } else {
+          generatedResource = {
+            url: buildUrl(resource.url, resource.params || {}),
+            json: true,
+            method: resource.method
+          };
+          if (resource.data) {
+            generatedResource.data = resource.data;
           }
         }
-      });
 
-      mySocket.send({
-        action: 'request',
-        resource: resource
-      });
+        self.bindings.push({
+          resource: generatedResource,
+          callback: cb,
+          resolve: resolve,
+          reject: reject
+        });
 
-      return deferred.promise;
+        mySocket.send({
+          action: 'request',
+          resource: generatedResource
+        });
+
+      }, false);
+      if (!resource.$isResource) {
+        prom = prom.then(function(res) {
+          res = res.data;
+          return res;
```

Comment:
This is so that it is compatible with the $resource object. It needs to return the value to be used by the subsequent ```then``` handler functions.